app/models/account/syncable.rb

```diff
@@ -8,6 +8,12 @@ def sync_later(start_date = nil)
   def sync(start_date = nil)
     update!(status: "syncing")
 
+    if self.foreign_currency? && !ExchangeRate.exchange_rates_provider.initialized?
+      logger.error("Cancelling sync of foreign account #{id}: No exchange rate provider ready")
+      update!(status: "error")
+      return
+    end
```

Comment:
I like the idea of canceling a sync based on a condition here, but I think a `foreign_currency?` account should still be able to sync regardless of whether the exchange rates are available.

The reason being—as long as all the transactions/valuations are in the same currency, we can still successfully build a historical balance graph in that currency.

If we updated this to `self.multi_currency?` it makes more sense to me as we would require exchange rates to build even the native historical graph.

The "rollup" is the point of abstraction where the exchange rates missing cause an issue I think.

Reply:
Depending on what exchange rates are available during a sync, there is a chance that an error is raised in the calculator here https://github.com/maybe-finance/maybe/blob/main/app/models/account/balance/calculator.rb#L54
which is why I went with short-circuiting the entire sync process. 

If we're ok with not raising errors here and just generating invalid balances (together with showing the UI messages) then I can change this

Reply:
Yeah this is a tricky one.  To me this feels like the calculator logic I wrote is more at fault here than anything.

Ultimately, there are 2 stages of currency conversion going on in this calculator:

1. Normalizing `multi_currency?` balances - if this fails, then we can't show anything for the account 
2. Converting an entire series of `foreign_currency?` balances - if this fails, we can't rollup the account, but we can still show a history graph in its native currency

We could either:

1. Loop through balances and for balances where there is no exchange rate, we don't generate it at all (leaving balance gaps)
2. Pre-fetch exchange rates for the known balances and if any exchange fails, we abort the entire conversion

The "all or nothing" approach in #2 seems a bit simpler here and in order to make that happen I think this calculator is what needs the refactor to decouple the balance logic from the exchange rate logic.  The two are intermingled right now.  For example:

https://github.com/maybe-finance/maybe/blob/6e59fdb369d18dcd1163e9088f272a716ace6acb/app/models/account/balance/calculator.rb#L76-L82